### PR TITLE
Outline polish

### PR DESCRIPTION
**Pull request by hunger**
Sits on top of #1977 !

Small patches making sure the outline will show symbols in the right order and adding an icon to structs in the outline widget.

### CODE REVIEW

File: tools/lsp/server_loop.rs [hunk 1]
```diff
@@ -1028,20 +1028,30 @@ fn get_document_symbols(
...
-                    children: gen_children(child, ds, make_range),
+                    children: gen_children(child, ds, offset_mapper),
                     ..ds.clone()
                 })
             })
             .collect::<Vec<_>>();
         (!r.is_empty()).then(|| r)
     }
 
+    r.sort_by(|a, b| {
```
Comment (ogoffart) [inline tools/lsp/server_loop.rs:1045]: Why is it not sorted in the first place? We should be traversing the children in order.

Comment (ogoffart) [reply tools/lsp/server_loop.rs:1045]: Ah, i see, indeed.

> I would really like to add more information like struct members and such :-) It's a bit tricky to find all the proper ranges to fill in though.

All the information should be in the done.  You can assume the node is a `syntax_nodes::ObjectType` and then you have the `ObjectTypeMember`  within it.

Why do you want this information though?
Now, i don't think we should have properties and members in the outline itself.

**Review by ogoffart — APPROVED** [review]

File: tools/lsp/server_loop.rs (hunk 1 above)
Comment (tronical) [reply tools/lsp/server_loop.rs:1045]: > Why do you want this information though? Now, i don't think we should have properties and members in the outline itself.

I agree with Olivier on this. The more information there is in the outline, the deeper it is, the harder it is to read and navigate. That is what I think the primary purpose of the outline is: A compass that gives you a view of the hierarchy of the UI and (by interactivity) a way to navigate it.

Comment (ogoffart) [reply tools/lsp/server_loop.rs:1045]: Right, the outline should contain the whole document :-)

> Yes, I noticed. The ranges are there, but you need to do some parsing to get the name/Type out.

You probably have a different definition of parsing than I do.

```rust
node.ObjectTypeMember().map(|f| {
   let name = parser::identifier_text(&f)?;
   let type = fields.get(name)?;
   let range = f.text_range();
   //...
})
```

**Comment by ogoffart** [comment]
(Note that vscode sorts the outline itself by range)

**Pull request merged**